File: rust/src/ntt_mont.rs
use crate::bigint::{next_power_of_two, BigUInt};

const MONT_MOD: u64 = 998244353;
const MONT_ROOT: u64 = 3;

const MONTGOMERY_R_INV: u64 = 17450252288407896063; // (-998244353)^-1 mod 2^64
const R2_MOD: u64 = 299560064;

#[inline]
fn montgomery_mul(a: u64, b: u64) -> u64 {
    let t = (a as u128) * (b as u128);
    let m = (t as u64).wrapping_mul(MONTGOMERY_R_INV);
    let m_mod = (m as u128) * (MONT_MOD as u128);
    let mut result = ((t + m_mod) >> 64) as u64;
    if result >= MONT_MOD {
        result -= MONT_MOD;
    }
    result
}

fn mod_pow_mont(mut base: u64, mut exp: u64) -> u64 {
    let mut result = montgomery_mul(1, R2_MOD);
    while exp > 0 {
        if exp & 1 == 1 {
            result = montgomery_mul(result, base);
        }
        base = montgomery_mul(base, base);
        exp >>= 1;
    }
    result
}

fn ntt_inplace(a: &mut [u64], root: u64, inverse: bool) {
    let n = a.len();

    let mut j = 0;
    for i in 1..n {
        let mut bit = n >> 1;
        while j & bit != 0 {
            j ^= bit;
            bit >>= 1;
        }
        j ^= bit;
        if i < j {
            a.swap(i, j);
        }
    }

    let mut len = 2;
    while len <= n {
        let wlen = mod_pow_mont(root, (MONT_MOD - 1) / len as u64);
        let wlen = if inverse {
            mod_pow_mont(wlen, MONT_MOD - 2)
        } else {
            wlen
        };

        for i in (0..n).step_by(len) {
            let mut w = montgomery_mul(1, R2_MOD);
            for j in 0..len / 2 {
                let u = a[i + j];
                let v = montgomery_mul(a[i + j + len / 2], w);
                a[i + j] = u + v;
                if a[i + j] >= MONT_MOD {
                    a[i + j] -= MONT_MOD;
                }
                a[i + j + len / 2] = u + MONT_MOD - v;
                if a[i + j + len / 2] >= MONT_MOD {
                    a[i + j + len / 2] -= MONT_MOD;
                }
                w = montgomery_mul(w, wlen);
            }
        }
        len <<= 1;
    }

    if inverse {
        let inv_n = mod_pow_mont(montgomery_mul(n as u64, R2_MOD), MONT_MOD - 2);
        for x in a.iter_mut() {
            *x = montgomery_mul(*x, inv_n);
        }
    }
}

pub fn biguint_mul_ntt_mont(a: &BigUInt, b: &BigUInt) -> BigUInt {
    if a.is_zero() || b.is_zero() {
        return BigUInt::new();
    }

    let n = next_power_of_two(a.words().len() + b.words().len() - 1);

    let mut fa: Vec<u64> = vec![0; n];
    let mut fb: Vec<u64> = vec![0; n];

    for i in 0..a.words().len() {
        fa[i] = montgomery_mul(a.words()[i] % MONT_MOD, R2_MOD);
    }
    for i in 0..b.words().len() {
        fb[i] = montgomery_mul(b.words()[i] % MONT_MOD, R2_MOD);
    }

    let mont_root = montgomery_mul(MONT_ROOT, R2_MOD);

    ntt_inplace(&mut fa, mont_root, false);
    ntt_inplace(&mut fb, mont_root, false);

    for i in 0..n {
        fa[i] = montgomery_mul(fa[i], fb[i]);
    }

    ntt_inplace(&mut fa, mont_root, true);

    let mut result = Vec::with_capacity(n);
    for i in 0..n {
        result.push(montgomery_mul(fa[i], 1));
    }

    BigUInt::from_slice(&result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ntt_small() {
        let a = BigUInt::from_u64(12345);
        let b = BigUInt::from_u64(67890);
        let c = biguint_mul_ntt_mont(&a, &b);
        let expected = BigUInt::from_u64(12345 * 67890);
        assert_eq!(c, expected);
    }
}

File: c/src/ntt_mont.c
#pragma once
#include <bigint.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <ntt_mont_math.h>


static void ntt_mont_inplace(uint64_t *a, size_t n, uint64_t mod, uint64_t root, bool inverse) {
    for (size_t i = 1, j = 0; i < n; i++) {
        size_t bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) {
            uint64_t t = a[i]; a[i] = a[j]; a[j] = t;
        }
    }
    
    uint64_t *roots = malloc(n * sizeof(uint64_t));
    roots[0] = 1;
    for (size_t len = 2; len <= n; len <<= 1) {
        uint64_t wlen = mod_pow(root, (mod - 1) / len, mod);
        if (inverse) wlen = mod_pow(wlen, mod - 2, mod);
        for (size_t i = len/2; i < len; i++) {
            roots[i] = mod_mul(roots[i - len/2], wlen, mod);
        }
    }
    
    size_t len = 2;
    while (len <= n) {
        for (size_t i = 0; i < n; i += len) {
            for (size_t j = 0; j < len / 2; j++) {
                uint64_t u = a[i + j];
                uint64_t v = mod_mul(a[i + j + len / 2], roots[j], mod);
                a[i + j] = u + v;
                if (a[i + j] >= mod) a[i + j] -= mod;
                a[i + j + len / 2] = u + mod - v;
                if (a[i + j + len / 2] >= mod) a[i + j + len / 2] -= mod;
            }
        }
        len <<= 1;
    }
    
    if (inverse) {
        uint64_t inv_n = mod_pow(n, mod - 2, mod);
        for (size_t i = 0; i < n; i++) {
            a[i] = mod_mul(a[i], inv_n, mod);
        }
    }
    
    free(roots);
}

BigUInt *biguint_mul_ntt_mont(const BigUInt *a, const BigUInt *b) {
    if (biguint_is_zero(a) || biguint_is_zero(b)) {
        return biguint_new();
    }
    
    const uint64_t MOD = 998244353;
    const uint64_t ROOT = 3;
    
    size_t n = 1;
    while (n < a->len + b->len - 1) n <<= 1;
    size_t result_len = a->len + b->len - 1;
    
    // Pre-allocate all vectors upfront (like Rust)
    uint64_t *fa = calloc(n, sizeof(uint64_t));
    uint64_t *fb = calloc(n, sizeof(uint64_t));
    
    if (!fa || !fb) {
        free(fa); free(fb);
        return NULL;
    }
    
    // Copy input words (modulo MOD)
    for (size_t i = 0; i < a->len; i++) fa[i] = a->words[i] % MOD;
    for (size_t i = 0; i < b->len; i++) fb[i] = b->words[i] % MOD;
    
    // NTT forward
    ntt_mont_inplace(fa, n, MOD, ROOT, false);
    ntt_mont_inplace(fb, n, MOD, ROOT, false);
    
    // Pointwise multiply
    for (size_t i = 0; i < n; i++) {
        fa[i] = mod_mul(fa[i], fb[i], MOD);
    }
    
    // Inverse NTT
    ntt_mont_inplace(fa, n, MOD, ROOT, true);
    
    // Pre-allocate result with exact size (like Rust Vec::with_capacity)
    BigUInt *res = biguint_new();
    if (!res) {
        free(fa); free(fb);
        return NULL;
    }
    
    // Pre-allocate result words array
    res->words = calloc(result_len, sizeof(uint64_t));
    if (!res->words) {
        free(fa); free(fb);
        free(res);
        return NULL;
    }
    
    // Copy result (only up to result_len)
    for (size_t i = 0; i < result_len; i++) {
        res->words[i] = fa[i];
    }
    res->len = result_len;
    res->cap = result_len;
    
    free(fa); free(fb);
    biguint_normalize(res);
    
    return res;
}
